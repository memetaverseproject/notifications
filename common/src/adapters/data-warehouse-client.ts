import { IConfigComponent, IFetchComponent, ILoggerComponent } from '@well-known-components/interfaces'

export type Event = {
  context: 'notification_server'
  event: 'email_sent' | 'email_validation_started' | 'email_validated' | 'subscription_changed' | 'user_unsubscribed'
  body: {
    env: 'prd' | 'dev'
  } & any
}

export type IDataWarehouseClient = {
  sendEvent: (event: Event) => Promise<void>
}

export type IDwhGenericEventsClientComponents = {
  config: IConfigComponent
  logs: ILoggerComponent
  fetch: IFetchComponent
}

export async function createDataWarehouseClient(
  components: Pick<IDwhGenericEventsClientComponents, 'config' | 'fetch' | 'logs'>
): Promise<IDataWarehouseClient> {
  const { config } = components

  const [apiBaseUrl, apiToken] = await Promise.all([config.getString('DWH_API_URL'), config.getString('DWH_TOKEN')])
  if (!apiBaseUrl || !apiToken) {
    return createDummyDataWarehouseClient(components)
  }

  return createDwhGenericEventsClient(components)
}

export async function createDwhGenericEventsClient(
  components: Pick<IDwhGenericEventsClientComponents, 'config' | 'fetch' | 'logs'>
): Promise<IDataWarehouseClient> {
  const { fetch, logs, config } = components
  const logger = logs.getLogger('dwh-generic-events-client')
  logger.info('Creating DWH generic events client')

  const [env, apiBaseUrl, apiToken] = await Promise.all([
    config.requireString('ENV'),
    config.requireString('DWH_API_URL'),
    config.requireString('DWH_TOKEN')
  ])

  async function sendEvent(event: Event): Promise<void> {
    logger.info(`Sending event to DataWarehouse ${event.event}"`)

    await fetch.fetch(apiBaseUrl, {
      method: 'POST',
      headers: {
        'x-token': apiToken,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        ...event,
        body: {
          ...event.body,
          env
        }
      })
    })
  }

  return {
    sendEvent
  }
}

export async function createDummyDataWarehouseClient(
  components: Pick<IDwhGenericEventsClientComponents, 'logs'>
): Promise<IDataWarehouseClient> {
  const { logs } = components
  const logger = logs.getLogger('dummy-data-warehouse-client')
  logger.info('Creating dummy DataWarehouse client')

  async function sendEvent(event: Event): Promise<void> {
    logger.debug(`Not sending event ${event.event}" to DataWarehouse`)
  }

  return {
    sendEvent
  }
}
