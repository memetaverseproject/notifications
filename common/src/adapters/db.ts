import SQL, { SQLStatement } from 'sql-template-strings'
import { NotificationDb, NotificationEvent, NotificationRecord, SubscriptionDb, UnconfirmedEmailDb } from '../types'
import { IPgComponent } from '@well-known-components/pg-component'
import { defaultSubscription } from '../subscriptions'
import { Email, EthAddress, NotificationChannelType, NotificationType, SubscriptionDetails } from '@mtvproject/schemas'

export type DbComponents = {
  pg: IPgComponent
}

export type UpsertResult<T> = {
  inserted: T[]
  updated: T[]
}

export type DbComponent = {
  findSubscription(address: EthAddress): Promise<SubscriptionDb>
  findSubscriptions(addresses: EthAddress[]): Promise<SubscriptionDb[]>
  findNotification(id: string): Promise<NotificationDb | undefined>
  findNotifications(users: EthAddress[], onlyUnread: boolean, from: number, limit: number): Promise<NotificationDb[]>
  markNotificationsAsRead(userId: EthAddress, notificationIds: string[]): Promise<number>
  saveSubscriptionDetails(address: EthAddress, subscriptionDetails: SubscriptionDetails): Promise<void>
  saveSubscriptionEmail(address: EthAddress, email: Email | undefined): Promise<void>
  findUnconfirmedEmail(address: EthAddress): Promise<UnconfirmedEmailDb | undefined>
  saveUnconfirmedEmail(address: EthAddress, email: string, code: string): Promise<void>
  deleteUnconfirmedEmail(address: EthAddress): Promise<void>
  fetchLastUpdateForNotificationType(notificationType: string): Promise<number>
  updateLastUpdateForNotificationType(notificationType: string, timestamp: number): Promise<void>
  insertNotifications(notificationRecord: NotificationRecord[]): Promise<UpsertResult<NotificationRecord>>
}

export function createDbComponent({ pg }: Pick<DbComponents, 'pg'>): DbComponent {
  async function findSubscription(address: EthAddress): Promise<SubscriptionDb> {
    return (await findSubscriptions([address]))[0]
  }

  async function findSubscriptions(addresses: EthAddress[]): Promise<SubscriptionDb[]> {
    const query: SQLStatement = SQL`
        SELECT address,
               email,
               details,
               created_at,
               updated_at
        FROM subscriptions n
        WHERE address = ANY (${addresses.map((a) => a.toLowerCase())})
    `

    const result = await pg.query<SubscriptionDb>(query)
    const indexedByAddress = result.rows.reduce(
      (acc, row) => {
        acc[row.address] = autoMigrate(row)
        return acc
      },
      {} as Record<EthAddress, SubscriptionDb>
    )

    return addresses.map(
      (address) =>
        indexedByAddress[address.toLowerCase()] || {
          address: address.toLowerCase(),
          email: undefined,
          details: defaultSubscription(),
          created_at: Date.now(),
          updated_at: Date.now()
        }
    )
  }

  async function findNotification(id: string): Promise<NotificationDb | undefined> {
    const query: SQLStatement = SQL`
        SELECT id,
               event_key,
               type,
               address,
               metadata,
               timestamp,
               read_at,
               created_at,
               updated_at
        FROM notifications
        WHERE id = ${id}
    `

    const result = await pg.query<NotificationDb>(query)
    if (result.rowCount === 0) {
      return undefined
    }

    return result.rows[0]
  }

  async function findNotifications(
    users: EthAddress[],
    onlyUnread: boolean,
    from: number,
    limit: number
  ): Promise<NotificationDb[]> {
    const query: SQLStatement = SQL`
        SELECT id,
               event_key,
               type,
               n.address as address,
               metadata,
               timestamp,
               n.read_at as read_at,
               created_at,
               updated_at,
               br.address AS broadcast_address,
               br.read_at AS broadcast_read_at
        FROM notifications n
        LEFT JOIN broadcast_read br ON n.id = br.notification_id
    `

    const lowercaseUsers = users.map((u) => u.toLowerCase())
    const whereClause: SQLStatement[] = [SQL`(n.address IS NULL OR n.address = ANY (${lowercaseUsers}))`]
    if (from > 0) {
      whereClause.push(SQL`timestamp >= ${from}`)
    }
    if (onlyUnread) {
      whereClause.push(SQL`(n.address IS NOT NULL AND n.read_at IS NULL) OR (n.address IS NULL AND br.read_at IS NULL)`)
    }
    let where = SQL` WHERE `.append(whereClause[0])
    for (const condition of whereClause.slice(1)) {
      where = where.append(' AND ').append(condition)
    }

    query.append(where)
    query.append(SQL` ORDER BY timestamp DESC`)
    query.append(SQL` LIMIT ${limit}`)

    return (await pg.query<NotificationDb>(query)).rows
  }

  async function markNotificationsAsRead(userId: EthAddress, notificationIds: string[]) {
    const readAt = Date.now()

    const updateResult = await pg.query<NotificationEvent>(SQL`
        UPDATE notifications
        SET    read_at    = ${readAt},
               updated_at = ${readAt}
        WHERE  read_at IS NULL
          AND  address = ${userId.toLowerCase()}
          AND  id = ANY (${notificationIds})
        RETURNING id
    `)
    let notificationCount = updateResult.rowCount

    const addressedNotificationsIds = new Set(updateResult.rows.map((n) => n.id))
    const potentialBroadcastIds = notificationIds.filter((id) => !addressedNotificationsIds.has(id))
    if (potentialBroadcastIds.length > 0) {
      const lowerUserId = userId.toLowerCase()
      const query = SQL`
        INSERT INTO broadcast_read (notification_id, address, read_at)
          SELECT id, ${lowerUserId}, ${readAt}
          FROM   notifications
          WHERE  id = ANY (${potentialBroadcastIds})
            AND  address IS NULL
        ON CONFLICT (notification_id, address) DO NOTHING`

      notificationCount += (await pg.query<NotificationEvent>(query)).rowCount
    }

    return notificationCount
  }

  async function saveSubscriptionDetails(address: string, subscriptionDetails: SubscriptionDetails): Promise<void> {
    const query: SQLStatement = SQL`
        INSERT INTO subscriptions (address, details, created_at, updated_at)
        VALUES (${address.toLowerCase()},
                ${subscriptionDetails}::jsonb,
                ${Date.now()},
                ${Date.now()}
        )
        ON CONFLICT (address) DO UPDATE
              SET details = ${subscriptionDetails}::jsonb,
              updated_at = ${Date.now()}
    `

    await pg.query(query)
  }

  async function saveSubscriptionEmail(address: string | undefined, email: Email | undefined): Promise<void> {
    const query: SQLStatement = SQL`
        INSERT INTO subscriptions (address, email, details, created_at, updated_at)
        VALUES (${address?.toLowerCase()},
                ${email},
                ${defaultSubscription()}::jsonb,
                ${Date.now()},
                ${Date.now()}
        )
        ON CONFLICT (address) DO UPDATE
              SET email = ${email},
              updated_at = ${Date.now()}
    `

    await pg.query(query)
  }

  async function findUnconfirmedEmail(address: string): Promise<UnconfirmedEmailDb | undefined> {
    const result = await pg.query<UnconfirmedEmailDb>(SQL`
        SELECT address, email, code, created_at, updated_at
        FROM unconfirmed_emails
        WHERE address = ${address.toLowerCase()};
    `)
    if (result.rowCount === 0) {
      return undefined
    }

    return result.rows[0]
  }

  async function saveUnconfirmedEmail(address: string, email: string, code: string): Promise<void> {
    const query: SQLStatement = SQL`
        INSERT INTO unconfirmed_emails (address, email, code, created_at, updated_at)
        VALUES (${address.toLowerCase()},
                ${email},
                ${code},
                ${Date.now()},
                ${Date.now()}
        )
        ON CONFLICT (address) DO UPDATE
              SET email = ${email},
                  code = ${code},
                  updated_at = ${Date.now()}
    `

    await pg.query(query)
  }

  async function deleteUnconfirmedEmail(address: string): Promise<void> {
    const query: SQLStatement = SQL`
        DELETE
        FROM unconfirmed_emails
        WHERE address = ${address.toLowerCase()}
    `

    await pg.query(query)
  }
  async function fetchLastUpdateForNotificationType(notificationType: string): Promise<number> {
    const result = await pg.query<{ last_successful_run_at: number }>(SQL`
        SELECT *
        FROM cursors
        WHERE id = ${notificationType};
    `)
    if (result.rowCount === 0) {
      return Date.now()
    }

    return result.rows[0].last_successful_run_at
  }

  async function updateLastUpdateForNotificationType(notificationType: string, timestamp: number): Promise<void> {
    const query = SQL`
        INSERT INTO cursors (id, last_successful_run_at, created_at, updated_at)
        VALUES (${notificationType}, ${timestamp}, ${Date.now()}, ${Date.now()})
        ON CONFLICT (id) DO UPDATE
        SET last_successful_run_at = ${timestamp},
            updated_at             = ${Date.now()};
    `

    await pg.query<any>(query)
  }

  async function insertNotifications(
    notificationRecords: NotificationRecord[]
  ): Promise<UpsertResult<NotificationRecord>> {
    const upsertResult: UpsertResult<NotificationRecord> = {
      inserted: [],
      updated: []
    }

    for (const notificationRecord of notificationRecords) {
      const buildQuery = SQL`
          INSERT INTO notifications (event_key, type, address, metadata, timestamp, read_at, created_at, updated_at)
          VALUES (${notificationRecord.eventKey},
                  ${notificationRecord.type},
                  ${notificationRecord.address?.toLowerCase() || null},
                  ${notificationRecord.metadata}::jsonb,
                  ${notificationRecord.timestamp},
                  NULL,
                  ${Date.now()},
                  ${Date.now()})
          ON CONFLICT (event_key, type, address) DO UPDATE
              SET metadata   = ${notificationRecord.metadata}::jsonb,
                  timestamp  = ${notificationRecord.timestamp},
                  updated_at = ${Date.now()}
          RETURNING id, xmax;
      `

      const result = await pg.query(buildQuery)
      notificationRecord.id = result.rows[0].id
      if (result.rows[0].xmax === '0') {
        upsertResult.inserted.push(notificationRecord)
      } else {
        upsertResult.updated.push(notificationRecord)
      }
    }

    return upsertResult
  }

  return {
    findNotification,
    findSubscription,
    findSubscriptions,
    fetchLastUpdateForNotificationType,
    updateLastUpdateForNotificationType,
    insertNotifications,
    findNotifications,
    markNotificationsAsRead,
    saveSubscriptionDetails,
    saveSubscriptionEmail,
    findUnconfirmedEmail,
    saveUnconfirmedEmail,
    deleteUnconfirmedEmail
  }
}

function autoMigrate(row: SubscriptionDb): SubscriptionDb {
  const defSubscription = defaultSubscription()
  const validMessageTypes = Object.keys(row.details.message_type)
    .filter((key) => key in defSubscription.message_type)
    .reduce(
      (obj, key) => {
        obj[key as NotificationType] = row.details.message_type[key as NotificationType]
        return obj
      },
      {} as Record<NotificationType, NotificationChannelType>
    )
  return {
    ...row,
    details: {
      ...defSubscription,
      ...row.details,
      message_type: {
        ...defSubscription.message_type,
        ...validMessageTypes
      }
    }
  }
}
