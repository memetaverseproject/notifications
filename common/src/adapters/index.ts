export * from './db'
export * from './sendgrid-client'
export * from './data-warehouse-client'
