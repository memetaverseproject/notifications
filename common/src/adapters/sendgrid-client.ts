import { IConfigComponent, IFetchComponent, ILoggerComponent } from '@well-known-components/interfaces'
import { Email } from '../types'

export type ISendGridClient = {
  sendEmail: (email: Email, customArgs?: CustomArgs) => Promise<void>
}

export type SendGridComponents = {
  config: IConfigComponent
  logs: ILoggerComponent
  fetch: IFetchComponent
}

export type EmailType = 'notification' | 'validation_attempt'

export type CustomArgs = {
  environment: string
  tracking_id?: string
  email_type?: EmailType
}

export async function createSendGrid(
  components: Pick<SendGridComponents, 'config' | 'fetch' | 'logs'>
): Promise<ISendGridClient> {
  const { config, fetch, logs } = components
  const logger = logs.getLogger('sendgrid-client')

  const [apiBaseUrl, apiKey, emailFrom, emailTemplateId, sandboxMode] = await Promise.all([
    config.requireString('SENDGRID_API_URL'),
    config.requireString('SENDGRID_API_KEY'),
    config.requireString('SENDGRID_EMAIL_FROM'),
    config.requireString('SENDGRID_EMAIL_TEMPLATE_ID'),
    config.getString('SENDGRID_SANDBOX_MODE')
  ])

  async function sendEmail(email: Email, customArgs?: CustomArgs): Promise<void> {
    logger.info(`Sending email to ${email.to} with subject "${email.subject}"`)
    const data = {
      personalizations: [
        {
          to: [
            {
              email: email.to
            }
          ],
          dynamic_template_data: {
            subject: email.subject,
            address: email.to,
            content: email.content,
            actionButtonLink: email.actionButtonLink,
            actionButtonText: email.actionButtonText,
            unsubscribeAllUrl: email.unsubscribeAllUrl,
            unsubscribeOneUrl: email.unsubscribeOneUrl
          }
        }
      ],
      from: {
        email: email.from || emailFrom,
        name: 'Memetaverse'
      },
      template_id: emailTemplateId,
      custom_args: customArgs,
      attachments: email.attachments,
      mail_settings: {
        sandbox_mode: {
          enable: sandboxMode === 'true'
        }
      }
    }

    await fetch.fetch(`${apiBaseUrl}/v3/mail/send`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data)
    })
  }

  return {
    sendEmail
  }
}
