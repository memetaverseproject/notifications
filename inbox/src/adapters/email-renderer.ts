import * as fs from 'node:fs'
import * as path from 'path'
import handlebars from 'handlebars'
import { Email } from '@notifications/common'

enum TemplatePart {
  SUBJECT = 'subject',
  CONTENT = 'content'
}

export enum InboxTemplates {
  VALIDATE_EMAIL = 'validate-email'
}

export type IEmailRenderer = {
  renderEmail(template: InboxTemplates, to: string, context: any): Promise<Email>
}

function loadTemplates() {
  return Object.values(InboxTemplates).reduce(
    (acc, template) => {
      acc[template] = {
        [TemplatePart.SUBJECT]: handlebars.compile(
          fs.readFileSync(
            path.join(__dirname, `email-templates/${template}.${TemplatePart.SUBJECT}.handlebars`),
            'utf8'
          )
        ),
        [TemplatePart.CONTENT]: handlebars.compile(
          fs.readFileSync(
            path.join(__dirname, `email-templates/${template}.${TemplatePart.CONTENT}.handlebars`),
            'utf8'
          )
        )
      }
      return acc
    },
    {} as Record<InboxTemplates, Record<TemplatePart, HandlebarsTemplateDelegate>>
  )
}

export async function createEmailRenderer(): Promise<IEmailRenderer> {
  const templates = loadTemplates()

  async function renderEmail(template: InboxTemplates, to: string, context: any): Promise<Email> {
    return {
      to,
      subject: templates[template][TemplatePart.SUBJECT](context),
      content: templates[template][TemplatePart.CONTENT](context)
    }
  }

  return {
    renderEmail
  }
}
