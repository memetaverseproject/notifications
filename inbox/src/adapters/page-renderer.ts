import * as fs from 'node:fs'
import * as path from 'path'
import handlebars from 'handlebars'

export type IPageRenderer = {
  renderPage(template: string, context: any): string
}

function loadTemplates() {
  const templatesFolder = path.join(__dirname, 'page-templates')
  return fs
    .readdirSync(templatesFolder, { encoding: 'utf8' })
    .filter((file) => file.endsWith('.handlebars'))
    .reduce(
      (acc, file) => {
        acc[file.replace('.handlebars', '')] = handlebars.compile(
          fs.readFileSync(path.join(templatesFolder, file), 'utf8')
        )
        return acc
      },
      {} as Record<string, HandlebarsTemplateDelegate>
    )
}

export async function createPageRenderer(): Promise<IPageRenderer> {
  const templates = loadTemplates()
  function renderPage(template: string, context: any): string {
    return templates[template](context)
  }

  return {
    renderPage
  }
}
