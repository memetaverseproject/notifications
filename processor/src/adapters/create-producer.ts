import { AppComponents, INotificationGenerator, INotificationProducer } from '../types'
import { CronJob } from 'cron'

export async function createProducer(
  components: Pick<AppComponents, 'logs' | 'db' | 'notificationsService'>,
  producer: INotificationGenerator
): Promise<INotificationProducer> {
  const { logs, db, notificationsService } = components
  const logger = logs.getLogger(`producer-${producer.notificationType}`)

  let lastSuccessfulRun: number | undefined

  async function runProducer(lastSuccessfulRun: number) {
    logger.info(`Checking for updates since ${lastSuccessfulRun}.`)

    const produced = await producer.run(lastSuccessfulRun)
    await notificationsService.saveNotifications(produced.records)
    await db.updateLastUpdateForNotificationType(produced.notificationType, produced.lastRun)
    logger.info(`Created ${produced.records.length} new notifications.`)
    return produced.lastRun
  }

  async function start(): Promise<void> {
    logger.info(`Scheduling producer for ${producer.notificationType}.`)

    const job = new CronJob(
      '0 * * * * *',
      async function () {
        try {
          if (!lastSuccessfulRun) {
            lastSuccessfulRun = await db.fetchLastUpdateForNotificationType(producer.notificationType)
          }
          lastSuccessfulRun = await runProducer(lastSuccessfulRun!)
        } catch (e: any) {
          logger.error(`Couldn't run producer: ${e.message}.`)
        }
      },
      null,
      false,
      'UCT'
    )
    job.start()
  }

  return {
    start,
    notificationType: () => producer.notificationType,
    runProducerSinceDate: async (date: number) => {
      await runProducer(date)
    }
  }
}
