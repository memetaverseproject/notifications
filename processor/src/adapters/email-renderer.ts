import * as fs from 'node:fs'
import * as path from 'path'
import handlebars from 'handlebars'
import { NotificationType } from '@mtvproject/schemas'
import { Email, NotificationRecord } from '@notifications/common'
import { AppComponents } from '../types'
import { formatMana } from '../logic/utils'
import { signUrl } from '@notifications/common/dist/signing'

export type IEmailRenderer = {
  renderTemplate(email: Email): Promise<string>
  renderEmail(emailAddress: string, notification: NotificationRecord): Promise<Email>
}

enum TemplatePart {
  SUBJECT = 'subject',
  CONTENT = 'content'
}

function loadTemplates() {
  handlebars.registerHelper('formatMana', (mana: string) => formatMana(mana))
  handlebars.registerHelper('escape', (variable: string) => {
    const s = JSON.stringify(variable || '""')
    // we remove the start and end quotes
    return s.substring(1, s.length - 1)
  })
  handlebars.registerHelper('land', (variable: string) => {
    // TODO let's do something better here
    return variable
  })
  handlebars.registerHelper('days', (from: string, to: string) => (parseInt(to) - parseInt(from)) / 86400)
  handlebars.registerHelper('insert', (text: string, defaultText: string) => (text ? text : defaultText))

  return Object.values(NotificationType).reduce(
    (acc, notificationType) => {
      acc[notificationType] = {
        [TemplatePart.SUBJECT]: handlebars.compile(
          fs.readFileSync(
            path.join(__dirname, `email-templates/${notificationType}.${TemplatePart.SUBJECT}.handlebars`),
            'utf8'
          ),
          { noEscape: true }
        ),
        [TemplatePart.CONTENT]: handlebars.compile(
          fs.readFileSync(
            path.join(__dirname, `email-templates/${notificationType}.${TemplatePart.CONTENT}.handlebars`),
            'utf8'
          )
        )
      }
      return acc
    },
    {} as Record<NotificationType, Record<TemplatePart, HandlebarsTemplateDelegate>>
  )
}

export async function createEmailRenderer(components: Pick<AppComponents, 'config'>): Promise<IEmailRenderer> {
  const [signingKey, serviceBaseUrl] = await Promise.all([
    components.config.requireString('SIGNING_KEY'),
    components.config.requireString('SERVICE_BASE_URL')
  ])

  const templates = loadTemplates()

  const emailTemplate = handlebars.compile(
    fs.readFileSync(path.join(__dirname, 'email-templates/email-template.handlebars'), 'utf8')
  )

  async function renderTemplate(email: Email): Promise<string> {
    return emailTemplate(email)
  }

  async function renderEmail(emailAddress: string, notification: NotificationRecord): Promise<Email> {
    const unsubscribeAllUrl = signUrl(
      signingKey,
      new URL(`/unsubscribe/${notification.address}`, serviceBaseUrl).toString()
    )
    const unsubscribeOneUrl = signUrl(
      signingKey,
      new URL(`/unsubscribe/${notification.address}/${notification.type}`, serviceBaseUrl).toString()
    )

    return {
      to: emailAddress,
      content: templates[notification.type][TemplatePart.CONTENT](notification),
      ...JSON.parse(templates[notification.type][TemplatePart.SUBJECT](notification)),
      unsubscribeAllUrl,
      unsubscribeOneUrl
    }
  }

  return {
    renderTemplate,
    renderEmail
  }
}
