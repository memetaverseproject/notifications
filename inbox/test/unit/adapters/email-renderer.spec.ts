import { createEmailRenderer, IEmailRenderer, InboxTemplates } from '../../../src/adapters/email-renderer'

describe('email rendering tests', () => {
  let renderer: IEmailRenderer

  type TestCaseData = {
    to: string
    context: any
  }

  const testCases: Record<InboxTemplates, TestCaseData> = {
    [InboxTemplates.VALIDATE_EMAIL]: {
      to: 'email@example.com',
      context: {
        validateButtonLink: 'https://memetaverse.club/account/confirm-email/0oekmfzAAJccU2d1el57WQWqJrb3FSuG',
        validateButtonText: 'Click Here to Confirm Your Email'
      }
    }
  }

  beforeAll(async () => {
    renderer = await createEmailRenderer()
  })

  const cases: [InboxTemplates, TestCaseData][] = Object.keys(testCases).map((type: InboxTemplates) => [
    type,
    testCases[type]
  ])

  test.each(cases)(`rendering %s`, async (type: InboxTemplates, data: TestCaseData) => {
    expect(await renderer.renderEmail(type, data.to, data.context)).toMatchSnapshot()
  })
})
