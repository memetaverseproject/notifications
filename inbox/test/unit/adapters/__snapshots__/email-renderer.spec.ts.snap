// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email rendering tests rendering validate-email 1`] = `
{
  "content": "<p>Hello!</p>
<p>You're receiving this email because you signed up for Memetaverse email notifications.</p>
<p>You'll be able to customize exactly what kind of notifications you'd like to receive (e.g., Marketplace, DAO, Events,
    etc.), but you must first confirm your email.</p>

<p style="height: 41px; padding-top: 8px; padding-bottom: 8px;"><a
        href="https://memetaverse.club/account/confirm-email/0oekmfzAAJccU2d1el57WQWqJrb3FSuG" target="_blank"
        rel="noopener"
        style="
            background-color: #ff2d55;
            width: 365px;
            height: 41px;
            border-radius: 6px;
            color: #fff;
            padding: 12px 44px 12px 44px;
            text-decoration: none;
            font-family: system-ui, BlinkMacSystemFont, 'Segoe UI',
              Roboto, 'Helvetica Neue', Arial, sans-serif;
          ">Click Here to Confirm Your Email</a>
</p>

<p>Or you can copy and paste this URL into the browser: <a href="https://memetaverse.club/account/confirm-email/0oekmfzAAJccU2d1el57WQWqJrb3FSuG" style="font-size: 13px;">https://memetaverse.club/account/confirm-email/0oekmfzAAJccU2d1el57WQWqJrb3FSuG</a>
</p>
",
  "subject": "Confirm your email",
  "to": "email@example.com",
}
`;
